kroxylicious/src/main/java/io/kroxylicious/proxy/KafkaProxy.java

```diff
@@ -103,96 +97,54 @@ public KafkaProxy startup() throws InterruptedException {
                 toBind.add(endpointProvider.getBrokerAddress(i));
             }
 
-            List<ChannelFuture> bindFutures = new ArrayList<>();
-            for (var address : toBind) {
-                var proxyPort = address.port();
-
-                if (endpointProvider.getBindAddress().isPresent()) {
-                    bindFuture = serverBootstrap.bind(endpointProvider.getBindAddress().get(), proxyPort);
-                }
-                else {
-                    bindFuture = serverBootstrap.bind(proxyPort);
-                }
-                bindFutures.add(bindFuture);
-
-                bindFuture.addListener(f -> {
-                    var channel = ((ChannelFuture) f).channel();
-                    acceptorChannels.add(channel);
-                    LOGGER.info("{}: Listener started {}", name, channel.localAddress());
-                });
-            }
-
-            bindFutures.forEach(ChannelFuture::syncUninterruptibly);
+            networkBinders.addAll(toBind.stream().map(hp -> NetworkBinding.createNetworkBinding(endpointProvider.getBindAddress(), hp.port(), virtualCluster.isUseTls()))
+                    .collect(Collectors.toSet()));
         });
 
+        var virtualHostEventGroup = buildNettyEventGroups(availableCores, false /* FIXME */);
+        this.virtualHostEventGroups.add(virtualHostEventGroup);
+
+        var tlsServerBootstrap = buildServerBootstrap(virtualHostEventGroup, new KafkaProxyInitializer(config, true, this, false, Map.of()));
+        var plainServerBootstrap = buildServerBootstrap(virtualHostEventGroup, new KafkaProxyInitializer(config, false, this, false, Map.of()));
+
+        var bindFutures = networkBinders.stream()
+                .map(networkBinder -> networkBinder.bind(networkBinder.tls() ? tlsServerBootstrap : plainServerBootstrap))
+                .collect(Collectors.toSet());
+        bindFutures.forEach(channelFuture -> channelFuture.addListener(f -> LOGGER.info("Listening : {}", channelFuture.channel().localAddress())));
+        bindFutures.forEach(ChannelFuture::syncUninterruptibly);
+
         // Pre-register counters/summaries to avoid creating them on first request and thus skewing the request latency
         // TODO add a virtual host tag to metrics
         Metrics.inboundDownstreamMessagesCounter();
         Metrics.inboundDownstreamDecodedMessagesCounter();
         return this;
     }
 
+    private ServerBootstrap buildServerBootstrap(EventGroupConfig virtualHostEventGroup, KafkaProxyInitializer kafkaProxyInitializer) {
+        ServerBootstrap serverBootstrap = new ServerBootstrap().group(virtualHostEventGroup.bossGroup(), virtualHostEventGroup.workerGroup())
+                .channel(virtualHostEventGroup.clazz())
+                .option(ChannelOption.SO_REUSEADDR, true)
+                .childHandler(kafkaProxyInitializer)
+                .childOption(ChannelOption.TCP_NODELAY, true);
```

Comment:
I removed AUTO_READ false.  In order to avoid needing to control reads during the TLS handshake (which now go one before the rest of the pipeline exists), I need it to be that way.  The remainder of the kroxy pipeline takes care of  controlling autoRead status, so I believe this change is safe.